contrib/mesos/pkg/queue/delay/indexed_delay_queue.go

```diff
@@ -0,0 +1,248 @@
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/pivotal-golang/clock"
+
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue"
+	"k8s.io/kubernetes/contrib/mesos/pkg/queue/priority"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// IndexedDelayQueue is a thread-safe, time-based queue with uniquely identified items.
+//
+// Adding or offering an item with an ID matching an item still in the queue will
+// invoke the configured SchedulingPolicy and supplied ValueReplacementPolicy to
+// determine the desired behavior.
+//
+// Items can only be popped after their event time has been reached.
+// Use Add to specify a delay duration and have the event time calculated.
+// Use Offer to specify a specific event time.
+type IndexedDelayQueue struct {
+	*Queue
+	// items mapped by UniqueID - item set must match queue contents
+	items            map[string]priority.Item
+	schedulingPolicy SchedulingPolicy
+	clock            clock.Clock
+}
+
+func NewIndexedDelayQueue(clock clock.Clock) *IndexedDelayQueue {
+	return &IndexedDelayQueue{
+		Queue: NewDelayQueue(clock),
+		items: map[string]priority.Item{},
+		clock: clock,
+	}
+}
+
+// Add inserts an uniquely identified item into the queue,
+// provided another item with the same ID is not already present.
+func (q *IndexedDelayQueue) Add(d UniqueDelayed, rp ValueReplacementPolicy) {
+	q.push(&indexedAddedItem{
+		Item: priority.NewItem(d, NewDelayedPriority(d, q.clock)),
+		id:   d.GetUID(),
+	}, rp)
+}
+
+// indexedAddedItem wraps an item that was added to the queue.
+type indexedAddedItem struct {
+	priority.Item
+	id string
+}
+
+func (di indexedAddedItem) GetUID() string {
+	return di.id
+}
+
+// Push puts the item back into the queue (with the same ID)
+func (di *indexedAddedItem) Push(queue heap.Interface) {
+	dq := queue.(*IndexedDelayQueue)
+	dq.push(di, KeepExisting)
+}
+
+func (q *IndexedDelayQueue) Offer(d UniqueScheduled, rp ValueReplacementPolicy) bool {
+	if p, ok := NewScheduledPriority(d); ok {
+		q.push(&indexedOfferedItem{
+			Item: priority.NewItem(d, p),
+			id:   d.GetUID(),
+		}, rp)
+		return true
+	}
+	return false
+}
+
+func (q *IndexedDelayQueue) push(newItem UniqueItem, rp ValueReplacementPolicy) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+	id := newItem.GetUID()
+	item, exists := q.items[id]
+	if !exists {
+		item = newItem
+		heap.Push(q.Queue, item)
+		q.items[id] = item
+	} else {
+		// replace existing item
+		newValue := rp.Value(item.Value(), newItem.Value())
+		oldPriority := item.Priority().(Priority)
+		newPriority := newItem.Priority().(Priority)
+		newPriority = q.schedulingPolicy.EventTime(oldPriority, newPriority)
+		item = priority.NewItem(newValue, newPriority)
+
+		switch i := newItem.(type) {
+		case *indexedAddedItem:
+			item = &indexedAddedItem{Item: item, id: i.id}
+		case *indexedOfferedItem:
+			item = &indexedOfferedItem{Item: item, id: i.id}
+		default:
+			panic(fmt.Sprintf("unsupported newItem type: %v", reflect.TypeOf(i)))
+		}
+
+		heap.Fix(q.Queue, item.Index())
```

Comment:
The old implementation modified the `item` that was retrieved from the map. This one doesn't and instead reassigns the `item` pointer to a new object, leaving the internal state of `Queue` untouched. Hence `q.Queue` will contain a reference to the `item` retrieved from the map.

This line tells the heap to `Fix` for an item index, calculated from new item state. The Queue, however, has the old item state. The next line adjusts the state of the map but the state of the Queue is never reconciled.